src/network/network_osm_reader.cpp

```diff
@@ -0,0 +1,361 @@
+// network_osm_reader.cpp -- function to read OSM data as network
+
+#include "network/network.hpp"
+#include "util/debug.hpp"
+
+#include <unordered_map>
+#include <unordered_set>
+
+#include <osmium/handler.hpp>
+#include <osmium/io/any_input.hpp>
+#include <osmium/osm/node.hpp>
+#include <osmium/osm/way.hpp>
+#include <osmium/visitor.hpp>
+
+
+using namespace FMM;
+using namespace FMM::CORE;
+using namespace FMM::MM;
+using namespace FMM::NETWORK;
+
+
+
+/** Helper class to filter out OSM ways that are not roads */
+class way_filter {
+protected:
+  /** key -- value pairs; if any of the keys is present with any
+   * of the corresponding values, this item is skipped */
+  std::unordered_map<std::string, std::unordered_set<std::string> > exclude_tags;
+  /** key -- value pairs; all of the keys here need to be present
+   * with any of the values given -- if no value given, the key
+   * has to be present with any value */
+  std::unordered_map<std::string, std::unordered_set<std::string> > include_tags;
+
+public:
+  /** test if the given way matches the filters */
+  bool operator () (const osmium::Way& way) const {
```

Comment:
Basically the logic is the following:

1. include_tags are checked: the way must have all the keys in include_tags. If no values are given (empty set), then any value is accepted at this stage. If a set of values are given, only those are accepted.
Note that in practice, this step only checks that the "highway" tag is present with any value. So we actually don't need a generalized implementation here and this step could be replaced with a simple check for the "highway" tag.
I'll suggest such a simplified implementation.

2. exclude_tags are checked: we want to exclude way with specific tag values (key-value pairs), e.g. highway=footway is not suitable for driving

In the generalized implementation, a tag can be in both include_tags and exclude_tags. For highway, the first step makes it sure that it is present (with any value) and then the second step filters out values that are not suitable for driving.

Reply:
Also, I agree to rename the function if that makes it's purpose more clear :)

Reply:
https://github.com/dkondor/fmm/commit/8756815fb57929570c2791dcbab5a1da063f55d7
I suggest this as a simplification of way_filter -- it eliminates include_tags, since that is only used for testing for the presence of the "highway" tag key.
(note: I have not pushed it to this branch yet -- you can decide if this is an improvement here)